### ISSUE TITLE
[question] Updating single requirement on a lockfile

### ISSUE THREAD

**Issue by ericriff** [conan-io/conan#8829]
- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

Hi all, 

I've been using conan for a while now listing my requirements using the recipe revision as a lock, in order to get reproducible builds.
E.g.
```
   def requirements(self)::
      self.requires("apriltag/3.1.3@pro1/stable#3188e5c421356dae58dfbbe19c7b9f7a")
      self.requires("rapidjson/1.1.0@pro1/stable#699ad3bc537a8fa509c2005ca1bbe577")
      ....
```

I recently started looking into using lockfiles to achieve true build reproducibility, but I'm having some issues handling the locks. We deal with a bunch of different hosts so I'm creating different lockfiles like this:
```
$ conan lock create conan/conanfile.py --lockfile-out=conan/locks/host1 -pr:b x86_64_gcc8 -pr:h host1
```
Now our `conanfile.py` is complex (it deals with over 40 packages) and it is modified collaboratively by a bunch of developers. Without using the lockfiles, if a dev wanted to update a package he/she would just have to update the corresponding line in the conanfile. I'm not sure what the right methodology would be now. One option would be to remove the existing lockfile and creating a new one, but that could lead to a lot of changes depending on the state of the developer's conan cache.
 
 Lets say I just want to update rapidjson on the lockfile (just the recipe revision or the version) and keep the rest of it as is, how should I achieve this?

**Comment by memsharded** [comment]
Hi @ericriff 

The challenging issue is that doing modifications to conanfiles, can produce completely different graphs, that can be completely incompatible with the provided lockfile. It is possible to either lock things or to let them change, but not both simultaneously.

At the moment, to achieve an "approximate" lock, what could be done is to use ``conan lock install <lockfile>`` to install an existing lockfile in a **clean** or empty local cache. Then, taking the ``conanfile.py`` with the changes, and creating a new lockfile using that cache as a starting point will use things in the cache as much as possible, maintaining the previously locked thing as much as possible, but when something changes, it will manage to get the new changes as well. 

We are working to improve this, but it will take time, because it will require a lot of changes to the codebase and arquitecture, not sure how far we we will manage to get.

**Comment by ericriff** [comment]
Thanks for the comments @memsharded. 
I feel like this really cripples the lockfiles usability since I can only see them being useful on a project not being actively developed. On an living project I would expect the dependencies to change every now and then but I would like to have reproducible builds in the meantime.
Let's say that eventually a new version of one of the dependencies is needed, I think like Conan should be able to update the lockfile to accommodate this. Of course it should update said package but also its dependencies. At least that is how it works on other tools that use lockfiles, like poetry (python package manager). The existing infrastructure (not following your workaround) would mean changing the whole lockfile, not only the required changes).
I would like to have commands like
1. `conan lock update path-to-conanfile.py` : Adds new packages to the lockfile, if there are any. Also removes packages if they have been removed from the conanfile.py. Should also check for changes on RREV locks, version bumps, username/channel changes, etc.
2. `conan lock update path-to-conanfile.py --all` : Updates the lockfile by checking the remote/s. This can be done with existing commands.
1. `conan lock update path-to-conanfile.py --packages myPackage` : Similar to `1` but ignores everything that is not listed after `--packages`.

Regardless, I understand the complexity of this.

**Comment by prvg-sso** [comment]
>  I think like Conan should be able to update the lockfile to accommodate this. Of course it should update said package but also its dependencies. At least that is how it works on other tools that use lockfiles, like poetry (python package manager). The existing infrastructure (not following your workaround) would mean changing the whole lockfile, not only the required changes).
> 

I'm maintaining an SDK with libraries built by conan, this is exactly the feature we're struggling with: How to update one third party library at a time, without rebuilding all of them.

**Comment by memsharded** [comment]
Hi @prvg-sso , @ericriff 

I have great news here. Conan 2.0 fixed this with a completely revamped lockfiles feature.
Now lockfiles have a ``conan lock add`` command that allows to modify 1 single reference at a time, with much better behavior and simpler UX. Furthermore, it is also possible to apply the lockfile to the ``conan create`` process of the dependency, let it update the lockfile automatically with the new version (but still locking its partial dependencies), and then apply the lockfile to the consumers with the updated reference without issues. Lockfiles in 2.0 can even be merged, and 1 single lockfile can manage the multiple binary configurations.

I am closing this ticket as solved in 2.0, please have a look and try it, looking forward to your feedback (please open new tickets for it). Many thanks!
